r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+            T: Borrow<SWProjective<P>>,
+    {
+        let (x, y, infinity) = match value_gen() {
+            Ok(ge) => {
+                let ge = ge.borrow().into_affine();
+                (Ok(ge.x), Ok(ge.y), Ok(ge.infinity))
+            },
+            _ => (
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+            ),
+        };
+
+        // Perform on-curve check.
+        let b = P::COEFF_B;
+        let a = P::COEFF_A;
+
+        let x = NonNativeFieldGadget::alloc(&mut cs.ns(|| "x"), || x)?;
+        let y = NonNativeFieldGadget::alloc(&mut cs.ns(|| "y"), || y)?;
+        let infinity = Boolean::alloc(&mut cs.ns(|| "infinity"), || infinity)?;
+
+        // Check that y^2 = x^3 + ax +b
+        // We do this by checking that y^2 - b = x * (x^2 +a)
+        // let x2 = x.square(&mut cs.ns(|| "x^2"))?;
+        // let y2 = y.square(&mut cs.ns(|| "y^2"))?;
+        let x2 = x.mul(cs.ns(|| "x^2"), &x)?;
+        let y2 = y.mul(cs.ns(|| "y^2"), &y)?;
+
+        let x2_plus_a = x2.add_constant(cs.ns(|| "x^2 + a"), &a)?;
+        let y2_minus_b = y2.add_constant(cs.ns(|| "y^2 - b"), &b.neg())?;
+
+        let x2_plus_a_times_x = x2_plus_a.mul(cs.ns(|| "(x^2 + a)*x"), &x)?;
+
+        x2_plus_a_times_x.conditional_enforce_equal(
+            cs.ns(|| "on curve check"),
+            &y2_minus_b,
+            &infinity.not()
+        )?;
+
+        Ok(Self::new(x, y, infinity))
+    }
+
+    #[inline]
+    fn alloc_without_check<FN, T, CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        value_gen: FN,
+    ) -> Result<Self, SynthesisError>
+        where
+            FN: FnOnce() -> Result<T, SynthesisError>,
+            T: Borrow<SWProjective<P>>,
+    {
+        let (x, y, infinity) = match value_gen() {
+            Ok(ge) => {
+                let ge = ge.borrow().into_affine();
+                (Ok(ge.x), Ok(ge.y), Ok(ge.infinity))
+            },
+            _ => (
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+            ),
+        };
+
+        let x = NonNativeFieldGadget::alloc(&mut cs.ns(|| "x"), || x)?;
+        let y = NonNativeFieldGadget::alloc(&mut cs.ns(|| "y"), || y)?;
+        let infinity = Boolean::alloc(&mut cs.ns(|| "infinity"), || infinity)?;
+
+        Ok(Self::new(x, y, infinity))
+    }
+
+    #[inline]
+    fn alloc_checked<FN, T, CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        value_gen: FN,
+    ) -> Result<Self, SynthesisError>
+        where
+            FN: FnOnce() -> Result<T, SynthesisError>,
+            T: Borrow<SWProjective<P>>,
+    {
+        let alloc_and_prime_order_check =
+            |mut cs: r1cs_core::Namespace<_, _>, value_gen: FN| -> Result<Self, SynthesisError> {
```

Comment:
Why declaring the definition of alloc_prime_order_check in such an abstract implicit manner?